src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import com.mojang.serialization.MapCodec;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import net.minecraft.core.Holder;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.block.Block;
+import net.neoforged.neoforge.common.NeoForgeMod;
+
+/**
+ * Ingredient that matches ItemStacks of blocks from a TagKey<Block>, useful in cases
+ * like "minecraft:convertable_to_mud" where there isn't an accompanying item tag
+ * <p>
+ * Notice: This should not be used as a replacement for the normal Item Tag Ingredient,
+ * This should only be used when there is no way an item tag can be used in your use case
+ */
+public class BlockTagIngredient implements ICustomIngredient {
+    public static final MapCodec<BlockTagIngredient> CODEC = TagKey.codec(Registries.BLOCK)
+            .xmap(BlockTagIngredient::new, BlockTagIngredient::getTag).fieldOf("tag");
+
+    protected final TagKey<Block> tag;
+
+    @Nullable
+    protected ItemStack[] itemStacks;
+
+    public BlockTagIngredient(TagKey<Block> tag) {
+        this.tag = tag;
+    }
+
+    protected void dissolve() {
```

Comment:
This shouldn't be necessary, caching this is already handled by Ingredient. But is that cache accessible from the custom ingredient? Maybe not...

Reply:
Given this list is used in `test`, it might still be an issue as this cache is never invalidated. I suppose its only a problem if someone statically stores a block tag ingredient though.